### ISSUE TITLE
enable seccomp by default - [merged]

### ISSUE THREAD

**Issue by sebastinas** [pwmt/zathura#476]
*On GitLab by @valoq on Jun 1, 2018, 17:24*

---

_Merges master -> develop_

After some month of using zathura in strict sandbox mode without discovering new bugs, I believe seccomp can be enabled by default in the next release.

Since the default sandbox value is set to "normal" which should not interfere with normal operation anyway, new bugs resulting from the use of seccomp will likely only affect thuse who have the sandbox value set to "strict".

Before new releases are published I would advise for a quick test in strict sandbox mode to discover new bugs resulting from new functionality or changed library behavior. I will try to support this as best I can and periodically test the development branch to make sure no new bugs are introduced through seccomp.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 10, 2018, 17:27*

---

added 1 commit

<ul><li>25d68e52 - fix openat permissions</li></ul>

[Compare with previous version](!14/diffs?diff_id=57&start_sha=85ac40b579e0bd815dda5ff3c5e1da5f280c600e)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 10, 2018, 17:30*

---

fix for openat

since the flag argument is not at the same positions as with open, the sandbox did not prevent opening files for writing when openat is used.

**Comment by sebastinas** [comment]
There are still some bits and pieces that need checks if strict sandbox mode is used. Examples include `sc_print`. Did you also check all occurrences where zathura `forks` and `execs`?

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 16, 2018, 16:48*

---

I tested all functionality I could find in the manual. There are still some warning messages on the terminal like about missing bookmark functionality which can be cleaned up (will do so) but no more crashes so far.

All forks take place before zathura_init (and therefore seccomp) is called and I did not find any paths that could cause trouble in this respect.

What exactly do you mean by sc_print? The print function is already covered.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 24:59*

---

added 1 commit

<ul><li>c8fee8dc - fix kill sig_0 error</li></ul>

[Compare with previous version](!14/diffs?diff_id=58&start_sha=25d68e52082e6c24aeb2d87e6476284bd3b84766)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 17:01*

---

added 1 commit

<ul><li>eca2380d - document ubuntu bug</li></ul>

[Compare with previous version](!14/diffs?diff_id=59&start_sha=c8fee8dc615e853cfab173f7187f4f415256ab71)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 17:10*

Commented on [zathura/seccomp-filters.c line 217](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..6f425c51631b6a9314a549455f3a00c3f597c59b#diff-22bb54b772f234b195303663d5732b3cR217)



---

After some extensive testing on Ubuntu 18.04, 18.10, debian stretch and debian buster (alpha3), it seems this syscall is only required by ubuntu systems. (bug only appears with the newest patch on the openat syscall)

Maybe we can check on the target system and only allow this syscall if it is compiled on ubuntu but I don't really like the idea.

**Comment by sebastinas** [comment]
So if some library on Ubuntu uses that to check if the process is allowed to send signals, shouldn't other signals follow?

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 18:40*

Commented on [zathura/seccomp-filters.c line 217](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..6f425c51631b6a9314a549455f3a00c3f597c59b#diff-22bb54b772f234b195303663d5732b3cR217)



---

When allowing the kill syscall (without argument contrains) there is no other signal used. Only sig_0 is used twice according to strace.

However when ":" is used to enter commands, the process wants to use the "socket" syscall as well, which I did not yet investigate further as I am not convinced this should be required.

**Comment by sebastinas** [comment]
Could that have something to do with atk or various GTK+ supported input methods?

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 20:34*

Commented on [zathura/seccomp-filters.c line 217](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..6f425c51631b6a9314a549455f3a00c3f597c59b#diff-22bb54b772f234b195303663d5732b3cR217)



---

It was the ibus-daemon! The kill syscall was targeted at it's process id and after changing the input method in the settings zathura works as expected.

The question remains how to deal with it.

> The Intelligent Input Bus (IBus, pronounced as I-Bus) is an input method (IM) framework for multilingual input in Unix-like operating-systems. 

Does zathura need this? Can we prevent it's use with zathura (in strict sandbox mode)?

**Comment by sebastinas** [comment]
It would be wrong of zathura to disable/break the user's input method.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 21:00*

Commented on [zathura/seccomp-filters.c line 217](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..6f425c51631b6a9314a549455f3a00c3f597c59b#diff-22bb54b772f234b195303663d5732b3cR217)



---

As it stands zathura (in strict sandbox mode) either needs several additional syscalls  and talk to an external services, which does not sound very secure imho, or we disable the input method in strict sandbox mode by unsetting GTK_IM_MODULE (in the context of zathuras process only).

Alternatively we could create another config option to disable it for better sandboxing and allow the additional syscalls and the communication with ibus, although I am uncertain how much this affects the process isolation.

If I am not mistaken the communication with ibus works through domain sockets and uses the "socket" and "connect" syscall. Blocking those syscalls in zathura (after they are used to connect X11 and before the document is parsed) is one of the biggest benefits in my opinion. If (local) socket communication is not blocked, sandbox escape is quite easy (e.g. using dbus).

Therefore I would argue not to allow those syscalls for effective sandboxing.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 22:44*

---

added 1 commit

<ul><li>6f425c51 - fix ibus bug</li></ul>

[Compare with previous version](!14/diffs?diff_id=60&start_sha=eca2380d10d6052fe3b6efabada87f0c1fc40f86)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jun 29, 2018, 22:57*

Commented on [zathura/zathura.c line 441](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..228c0f249d564005acb0d239dbe3ccb4b5ec2b30#diff-46b90f7beb3ac107e23663e4c5bb9ec0L441)



---

This will prevent interference from input method frameworks like ibus.

Unsetting "GTK_IM_MODULE" like this will only disable the use of such frameworks within the zathura process context and only if strict sandbox mode is used.


I believe this to be the correct fix due to the following reasons:

allowing communication to services like ibus does:
- require several more syscalls
- allow communication to external services via ipc/domain sockets (basically sandbox escape)
- require different syscalls and communication methods as well as extensive tests for all alternative input method frameworks https://en.wikipedia.org/wiki/List_of_input_methods_for_UNIX_platforms
- break zathura/seccomp support again as soon as a new/untested service is used

**Comment by sebastinas** [comment]
Why unset the environment variable at all? If the kill syscall fails and GTK+ cannot communicate with the ibus daemon, wouldn't this be good enough?

**Comment by sebastinas** [comment]
Please start to document all the deficiencies of strict sandboxing mode in the manpages.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 4, 2018, 21:44*

Commented on [zathura/zathura.c line 441](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..228c0f249d564005acb0d239dbe3ccb4b5ec2b30#diff-46b90f7beb3ac107e23663e4c5bb9ec0L441)



---

If the syscall fails because it it not white listed the process is terminated and zathura crashes. I do not know of any alternative way to prevent the communication and therefore the syscall itself.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 4, 2018, 21:45*

Commented on [zathura/seccomp-filters.c line 217](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..6f425c51631b6a9314a549455f3a00c3f597c59b#diff-22bb54b772f234b195303663d5732b3cR217)



---

Will do

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 4, 2018, 21:45*

---

resolved all discussions

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 8, 2018, 11:13*

Commented on [zathura/seccomp-filters.c line 217](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..6f425c51631b6a9314a549455f3a00c3f597c59b#diff-22bb54b772f234b195303663d5732b3cR217)



---

changed this line in [version 6 of the diff](!14/diffs?diff_id=61&start_sha=6f425c51631b6a9314a549455f3a00c3f597c59b#74b58f99ea157b9c1c052e81ec83f8b18c1e8241_217_210)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 8, 2018, 11:13*

---

added 2 commits

<ul><li>52d06598 - Revert &quot;document ubuntu bug&quot;</li><li>228c0f24 - Revert &quot;fix kill sig_0 error&quot;</li></ul>

[Compare with previous version](!14/diffs?diff_id=61&start_sha=6f425c51631b6a9314a549455f3a00c3f597c59b)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 8, 2018, 12:10*

Commented on [zathura/zathura.c line 441](https://github.com/pwmt/zathura/compare/a4ffad457c35a2b39b38414eec830f06fab772ad..228c0f249d564005acb0d239dbe3ccb4b5ec2b30#diff-46b90f7beb3ac107e23663e4c5bb9ec0L441)



---

changed this line in version 6 of the diff

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 8, 2018, 12:12*

---

added 36 commits

<ul><li>68eaadf1...2a48d2df - 36 commits from branch <code>pwmt:develop</code></li></ul>

[Compare with previous version](!14/diffs?diff_id=63&start_sha=68eaadf1304630eff77aedad9170676533092d7f)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 8, 2018, 12:18*

---

added 1 commit

<ul><li>0e5dc9f2 - seccomp fixes</li></ul>

[Compare with previous version](!14/diffs?diff_id=64&start_sha=2a48d2df0f8d25cda6fe81b85df9e7d961aeeda9)

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Jul 8, 2018, 13:09*

---

added 1 commit

<ul><li>cb6922f7 - document features blocked by strict sandbox mode</li></ul>

[Compare with previous version](!14/diffs?diff_id=65&start_sha=0e5dc9f284a417fd560498466908ce8cdc2114db)

**Comment by sebastinas** [comment]
merged
